runtime/compiler/runtime/JITServerAOTCache.cpp

```diff
@@ -0,0 +1,683 @@
+         _name.c_str(), record->data().id(), (unsigned long long)clientUID
+      );
+
+   return record;
+   }
+
+
+bool
+JITServerAOTCache::storeMethod(const AOTCacheClassChainRecord *definingClassChainRecord, uint32_t index,
+                               TR_Hotness optLevel, const AOTCacheAOTHeaderRecord *aotHeaderRecord,
+                               const Vector<std::pair<const AOTCacheRecord *, uintptr_t/*reloDataOffset*/>> &records,
+                               const void *code, size_t codeSize, const void *data, size_t dataSize,
+                               const char *signature, uint64_t clientUID)
+   {
+   uintptr_t definingClassId = definingClassChainRecord->records()[0]->data().id();
+   const char *levelName = TR::Compilation::getHotnessName(optLevel);
+
+   CachedMethodKey key(definingClassChainRecord, index, optLevel, aotHeaderRecord);
+   OMR::CriticalSection cs(_cachedMethodMonitor);
+
+   auto it = _cachedMethodMap.find(key);
+   if (it != _cachedMethodMap.end())
+      {
+      //NOTE: Current implementation keeps the first version of the method for this key in the cache.
+      //      If we want to keep the most recent version instead, we will need to synchronize deleting
+      //      the old version with any concurrent threads that could be sending it to other clients.
+      if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+         TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
+            "AOT cache %s: method %s @ %s index %u class ID %zu AOT header ID %zu already exists",
+            _name.c_str(), signature, levelName, index, definingClassId, aotHeaderRecord->data().id()
+         );
+      return false;
+      }
+
+   auto method = CachedAOTMethod::create(definingClassChainRecord, index, optLevel, aotHeaderRecord,
+                                         records, code, codeSize, data, dataSize);
+   addToMap(_cachedMethodMap, it, key, method);
+
+   if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
+         "AOT cache %s: stored method %s @ %s index %u class ID %zu AOT header ID %zu for clientUID %llu",
+         _name.c_str(), signature, levelName, index, definingClassId,
+         aotHeaderRecord->data().id(), (unsigned long long)clientUID
+      );
+
+   return true;
+   }
+
+const CachedAOTMethod *
+JITServerAOTCache::findMethod(const AOTCacheClassChainRecord *definingClassChainRecord, uint32_t index,
+                              TR_Hotness optLevel, const AOTCacheAOTHeaderRecord *aotHeaderRecord)
+   {
+   CachedMethodKey key(definingClassChainRecord, index, optLevel, aotHeaderRecord);
+   OMR::CriticalSection cs(_cachedMethodMonitor);
+
+   auto it = _cachedMethodMap.find(key);
+   if (it == _cachedMethodMap.end())
+      return NULL;
+
+   return it->second;
+   }
+
+
+Vector<const AOTSerializationRecord *>
+JITServerAOTCache::getSerializationRecords(const CachedAOTMethod *method, const KnownIdSet &knownIds,
```

Comment:
Some of the methods may benefits from comments explaining their purpose and meaning of parameters.
For instance, is this method executed by the client or server?

Reply:
The whole class `JITServerAOTCache` is only used at the server (I'll add a comment saying that). I'll add comments for the public methods of this class describing what they do and what their parameters are.